### ISSUE TITLE
optimization strategies

### ISSUE THREAD

**Issue by dmedine** [RhythmsOfRelating/HybridHarmony#15]
Following up on https://github.com/RhythmsOfRelating/HybridHarmony/issues/12#issuecomment-848130896

> 1. are there ways to optimize the matrix computation in correlation?

I think that the actual computations themselves aren't going to get much faster. I don't think we can do much better than numpy. The problem, as I see it, is more to do with memory allocation and garbage collection. The way it is set up now, each stage in the chain needs to get a copy of whatever data it is processing. So two copies are made here: https://github.com/RhythmsOfRelating/HybridHarmony/blob/master/src/analysis/correlation_perFreq.py#L195-L196, some allocation is happening here: https://github.com/RhythmsOfRelating/HybridHarmony/blob/master/src/analysis/correlation_perFreq.py#L207, a new array gets allocated in the next line (https://github.com/RhythmsOfRelating/HybridHarmony/blob/master/src/analysis/correlation_perFreq.py#L208), and then two more get created here: https://github.com/RhythmsOfRelating/HybridHarmony/blob/master/src/analysis/correlation_perFreq.py#L217-L218. I confess that I am not a great expert on handling this problem in Python, (in C, you could just pre-allocate all the memory you need and reuse it) but I do know that all this allocation is taking up time and the garbage collection is going to be working overtime as well.

> 2. is it helpful if we make analysis a QThread instead of a python built-in thread? That way, we can implement real-time normalization using signals/slots (e.g. analysis object reads input from main_GUI whenever the slider moves).

I don't see any real performance benefit here, but I think it is an excellent idea. It will definitely make the program smoother.

Another thing that Michele mentioned is that due to the background thread being so heavy, and the GUI thread being so greedy, on a single-core machine, this application is very slow. However, I am not too concerned with that. I didn't even know anyone had a single-core machine anymore ;-)

For now, I think we have more pressing issues to deal with. I am also going to move point 2 into a separate issue. That is a pretty well defined (if complex) enhancement to do.

**Comment by phoebsc** [comment]
For the allocation problem, can I just condense the functions? For example, I can change [this line](https://github.com/RhythmsOfRelating/HybridHarmony/blob/6232c6fafa5bb50261cea449f3c3e0bfb92135ba/src/analysis/correlation_perFreq.py#L218)
```
all_analytic = np.array([self._calculate_analytic(all_analytic, coeff) for c, coeff in enumerate(self.COEFFICIENTS)])
```
into
```
all_analytic = np.array([hilbert(lfilter(coeff[0], coeff[1], all_analytic)) for c, coeff in enumerate(self.COEFFICIENTS)])
```

such that [this function ](https://github.com/RhythmsOfRelating/HybridHarmony/blob/6232c6fafa5bb50261cea449f3c3e0bfb92135ba/src/analysis/correlation_perFreq.py#L200) is not needed anymore.

**Comment by dmedine** [comment]
Yes! I believe that should work. I am not sure if will improve performance, however, because I don't know what is actually happening in terms of memory allocation in `lfilter` and` hilbert`, but it might! Can you try both versions and see if there is any CPU drop? I think there will be some small performance increase since there is one less function symbol to locate and load onto the chip, but that will probably be negligible. Then again, I am so ignorant of how Python works under the hood, I really don't know what I am talking about. In C/C++, the compiler will optimize little things like this for you automatically.

**Comment by phoebsc** [comment]
I tried this and CPU didn't really drop (still around 25%). I'm making this change anyway, tho.

**Comment by dmedine** [comment]
Hmm. As I suspected. It does make the code much clearer, though. Good to keep the change.
